### PR TITLE
feat: support batching for torchObjectDetectionSegmenter

### PR DESCRIPTION
**Pull request by Kelton8Z**

### CODE REVIEW

File: segmenters/image/TorchObjectDetectionSegmenter/__init__.py [hunk 1]
```diff
@@ -99,25 +110,43 @@ def segment(self, blob: 'np.ndarray', *args, **kwargs) -> List[Dict]:
...
-        raw_img = np.copy(blob)
+        raw_img = np.copy(blob) # (2, 681, 1264, 3) when
+        # "Ensure the color channel axis is the default axis." i.e. c comes first
+        # e.g. (h,w,c) -> (c,h,w) / (n,h,w,c) -> (c,n,h,w)
+
         raw_img = _move_channel_axis(raw_img, self.channel_axis, self._default_channel_axis)
-        bboxes, scores, labels = self._predict(raw_img)
-        img = _load_image(raw_img * 255, self._default_channel_axis)
+        # raw_img = np.transpose(raw_img,[0,3,1,2]) # (n,h,w,c) -> (n,c,h,w)
+        print(f"raw img n: {raw_img.shape[1]}")
```
Comment (nan-wang) [inline segmenters/image/TorchObjectDetectionSegmenter/__init__.py:119]:
```suggestion
```

File: segmenters/image/TorchObjectDetectionSegmenter/__init__.py [hunk 2]
```diff
@@ -99,25 +110,43 @@ def segment(self, blob: 'np.ndarray', *args, **kwargs) -> List[Dict]:
...
+        batchedImgArr = np.asarray([raw_img[:, 0, :, :], raw_img[:, 1, :, :]])#raw_img[np.arange(raw_img.shape[1])[:,None,None], x,y]
```
Comment (nan-wang) [inline segmenters/image/TorchObjectDetectionSegmenter/__init__.py:120]:
```suggestion
        batchedImgArr = np.asarray([raw_img[:, 0, :, :], raw_img[:, 1, :, :]])
```

**Review by nan-wang — DISMISSED** [review]

File: segmenters/image/TorchObjectDetectionSegmenter/__init__.py [hunk 3]
```diff
@@ -99,25 +110,45 @@ def segment(self, blob: 'np.ndarray', *args, **kwargs) -> List[Dict]:
         :param args:  Additional positional arguments
         :param kwargs: Additional keyword arguments
         """
-        raw_img = np.copy(blob)
+        raw_img = np.copy(blob) # (2, 681, 1264, 3) with imgs/cars.jpg
+        # "Ensure the color channel axis is the default axis." i.e. c comes first
+        # e.g. (h,w,c) -> (c,h,w) / (n,h,w,c) -> (c,n,h,w)
```
Comment (JoanFM) [inline segmenters/image/TorchObjectDetectionSegmenter/__init__.py:115]: color needs to come second if there is batching involved

File: segmenters/image/TorchObjectDetectionSegmenter/__init__.py [hunk 4]
```diff
@@ -99,25 +110,45 @@ def segment(self, blob: 'np.ndarray', *args, **kwargs) -> List[Dict]:
...
-                # at this point, raw_img has the channel axis at the default tensor one
-                _img, top, left = _crop_image(img, target_size=target_size, top=top, left=left, how='precise')
-                _img = _move_channel_axis(np.asarray(_img).astype('float32'), -1, self.channel_axis)
-                label_name = self.label_name_map[label]
-                self.logger.debug(
-                    f'detected {label_name} with confidence {score} at position {(top, left)} and size {target_size}')
-                result.append(
-                    dict(offset=0, weight=1., blob=_img,
-                         location=(top, left), tags={'label': label_name}))
+        for i in range(len(batchedPredictions)):
```
Comment (JoanFM) [inline segmenters/image/TorchObjectDetectionSegmenter/__init__.py:123]: use more pythonic iteration

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: segmenters/image/TorchObjectDetectionSegmenter/__init__.py (hunk 3 above)
Comment (Kelton8Z) [reply segmenters/image/TorchObjectDetectionSegmenter/__init__.py:103]: U mean as a convention ?

File: segmenters/image/TorchObjectDetectionSegmenter/__init__.py [hunk 5]
```diff
@@ -99,25 +98,44 @@ def segment(self, blob: 'np.ndarray', *args, **kwargs) -> List[Dict]:
...
         """
-        raw_img = np.copy(blob)
+        raw_img = np.copy(blob) # (2, 681, 1264, 3) with imgs/cars.jpg
+        # "Ensure the color channel axis is the default axis." i.e. c comes first
+        # e.g. (h,w,c) -> (c,h,w) / (n,h,w,c) -> (c,n,h,w)
         raw_img = _move_channel_axis(raw_img, self.channel_axis, self._default_channel_axis)
-        bboxes, scores, labels = self._predict(raw_img)
-        img = _load_image(raw_img * 255, self._default_channel_axis)
+
+        batchedImgArr = np.asarray([raw_img[:, i, :, :] for i in range(raw_img.shape[1])])
```
Comment (JoanFM) [inline segmenters/image/TorchObjectDetectionSegmenter/__init__.py:106]: raw_img should already be batched

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: segmenters/image/TorchObjectDetectionSegmenter/manifest.yml [hunk 6]
```diff
@@ -1,4 +1,4 @@
-manifest_version: 1
+manifest_version: 2
```
Comment (JoanFM) [inline segmenters/image/TorchObjectDetectionSegmenter/manifest.yml:1]: this should not change, what needs to be updated the version of the image, not the manifest

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: segmenters/image/TorchObjectDetectionSegmenter/__init__.py (hunk 1 above)
Comment (Kelton8Z) [reply segmenters/image/TorchObjectDetectionSegmenter/__init__.py]: Deleted as suggested

File: segmenters/image/TorchObjectDetectionSegmenter/__init__.py (hunk 2 above)
Comment (Kelton8Z) [reply segmenters/image/TorchObjectDetectionSegmenter/__init__.py]: Deleted

File: segmenters/image/TorchObjectDetectionSegmenter/__init__.py [hunk 7]
```diff
@@ -2,7 +2,7 @@
 
 import numpy as np
 
-from jina.executors.decorators import single
+from jina.executors.decorators import single,batching
```
Comment (JoanFM) [inline segmenters/image/TorchObjectDetectionSegmenter/__init__.py:5]: remove `single` decorator

**Review by JoanFM — CHANGES_REQUESTED** [review]
Hey @Kelton8Z ,

We need to have some tests to see how it works with a random image where no detection will be found.

**Comment by JoanFM** [comment]
> Hey @Kelton8Z ,
> 
> We need to have some tests to see how it works with a random image where no detection will be found.

My bad, there are already such tests

**Review by JoanFM — APPROVED** [review]

**Pull request merged**